### PR TITLE
Adding backpressure support to NettyRequest

### PR DESCRIPTION
**Pull request by vgkholla**
This change adds backpressure support to `NettyRequest` allowing it to suspend (and resume) reading from the client based on a configurable watermark. This helps keep memory usage at the frontend in check.

This change also adds a test for the corner case where the underlying buffer from Netty's `ByteBuf` cannot be extracted and the data has to be copied.

Built and formatted.

**Primary reviewers : @pnarayanan, @xiahome**
**Expected time to review: 45 - 60 mins**

Adresses #299.  

**Unit test coverage:**

Class | Class, % | Method, % | Line, %
------- | ------------ | -------------- | ----------
NettyRequest | 100% (3/ 3) | 96.4% (27/ 28) | 98.7% (228/ 231)
NettyMultipartRequest | 100% (1/ 1) | 100% (6/ 6) | 92.9% (79/ 85)

### CODE REVIEW

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java [hunk 1]
```diff
@@ -82,5 +92,6 @@ public NettyConfig(VerifiableProperties verifiableProperties) {
     nettyServerMaxInitialLineLength = verifiableProperties.getInt("netty.server.max.initial.line.length", 4096);
     nettyServerMaxHeaderSize = verifiableProperties.getInt("netty.server.max.header.size", 8192);
     nettyServerMaxChunkSize = verifiableProperties.getInt("netty.server.max.chunk.size", 8192);
+    nettyServerRequestBufferWatermark = verifiableProperties.getInt("netty.server.request.buffer.watermark", 33554432);
```
Comment (pnarayanan) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: `32 * 1024 * 1024` here as well?

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java [hunk 2]
```diff
@@ -73,6 +73,16 @@
   @Default("8192")
   public final int nettyServerMaxChunkSize;
 
+  /**
+   * The threshold of the size of buffered data at which reading from a client channel will be suspended. If the size
+   * drops below the threshold, reading will be resumed. This value is respected on a per-request basis.
+   * Note that the actual amount of data buffered will be >= this number.
```
Comment (pnarayanan) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: "will be" or "may be"?

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java [hunk 3]
```diff
@@ -758,19 +983,53 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
...
 
   /**
+   * Splits the given {@code contentBytes} into {@code numChunks} chunks and stores them in {@code httpContents}.
+   * @param contentBytes the content that needs to be split.
+   * @param numChunks the number of chunks to split {@code contentBytes} into.
+   * @param httpContents the {@link List<HttpContent>} that will contain all the content in parts.
+   * @param useCopyForcingByteBuf if {@code true}, uses {@link CopyForcingByteBuf} instead of the default
+   *                              {@link ByteBuf}.
+   */
+  private void splitContent(byte[] contentBytes, int numChunks, List<HttpContent> httpContents,
```
Comment (xiahome) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: Should also test single full response case?

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java (hunk 2 above)
Comment (xiahome) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: This seems to be per channel. Any comment on if the number could be too large?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 4]
```diff
@@ -488,91 +523,106 @@ private void validateState(HttpContent httpContent)
...
-  public ContentWriteCallback(HttpContent httpContent, boolean isLast, ReadIntoCallbackWrapper callbackWrapper) {
-    this.httpContent = httpContent;
-    this.isLast = isLast;
-    this.callbackWrapper = callbackWrapper;
+  private void continueReadIfPossible(long delta) {
+    if (!channel.config().isAutoRead()) {
+      if (bytesBuffered.addAndGet(delta) < bufferWatermark) {
+        channel.read();
+      } else {
+        nettyMetrics.watermarkOverflowCount.inc();
```
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Would this be a highly-frequent occurrence in the situation data come in faster than being processed?

**Comment by xiahome** [comment]
Mostly looks good, will come to you for the back pressure test.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java [hunk 5]
```diff
@@ -749,7 +839,142 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
...
+        queuedReads.set(1);
+        bytesToVerify = 0;
+        suspended = false;
+      } else {
+        assertEquals("There should have been only one read queued", 1, queuedReads.get());
+        queuedReads.set(0);
+        if (future == null && addedCount == numChunksToAddBeforeRead) {
+          future = nettyRequest.readInto(writeChannel, callback);
+        }
+        if (addedCount < httpContents.size()) {
```
Comment (xiahome) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: This check seems to be unnecessary.

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java (hunk 1 above)
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: will change.

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java (hunk 2 above)
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: @pnarayanan "may be" sounds better will change
@xiahome I think this is fine - and since this is a configuration, can always be tuned.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 4 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Yes, it could be. But I want to see how this metric behaves and its relation to latencies etc.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 5 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: will remove.

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java (hunk 1 above)
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: Done.

File: ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java (hunk 2 above)
Comment (vgkholla) [inline ambry-api/src/main/java/com.github.ambry/config/NettyConfig.java]: Done.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 3 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: I'm assuming you mean full request i.e `FullHttpRequest` that contains both the request and content. That test is done as a part of `NettyMessageProcessorTest::rawBytesPostTest()` where it belongs.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 5 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: Done.

**Comment by vgkholla** [comment]
@pnarayanan @xiahome addressed your comments

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 6]
```diff
@@ -52,9 +53,14 @@
  * A wrapper over {@link HttpRequest} and all the {@link HttpContent} associated with the request.
  */
 class NettyRequest implements RestRequest {
+  // If the write of at least {@code bufferWatermark} amount of data is unacknowledged, reading from the channel will be
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: do you want to highlight that this is for PUTs only?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMessageProcessor.java [hunk 7]
```diff
@@ -283,11 +284,11 @@ private boolean handleRequest(HttpRequest httpRequest)
         try {
           // We need to maintain state about the request itself for the subsequent parts (if any) that come in. We will
           // attach content to the request as the content arrives.
-          if (HttpPostRequestDecoder.isMultipart(httpRequest)) {
+          if (HttpMethod.POST.equals(httpRequest.getMethod()) && HttpPostRequestDecoder.isMultipart(httpRequest)) {
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMessageProcessor.java]: why add this check now? And is this to guard against a Multipart non-POST method?

Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMessageProcessor.java]: And I think there is a redundant check within `NettyMultipartRequest` constructor.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 8]
```diff
@@ -52,9 +53,14 @@
...
+  // temporarily suspended. It will be resumed when the amount of data unacknowledged drops below this number. If this
+  // is <=0, it is assumed that there is no limit on the size of unacknowledged data.
+  static int bufferWatermark = -1;
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: is this variable really required? does it help with the tests?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 9]
```diff
@@ -92,34 +100,42 @@
...
-    if (request == null) {
-      throw new IllegalArgumentException("Received null HttpRequest");
+    if (request == null || channel == null) {
+      throw new IllegalArgumentException("Received null argument(s)");
     }
     restRequestMetricsTracker.nioMetricsTracker.markRequestReceived();
+    this.request = request;
+    query = new QueryStringDecoder(request.getUri());
+    this.channel = channel;
+    savedMaxMessagesPerRead = channel.config().getMaxMessagesPerRead();
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: what is a message? An `HttpContent`?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 8 above)
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: and is there a practical difference between 0 and -1 here?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 10]
```diff
@@ -488,91 +523,106 @@ private void validateState(HttpContent httpContent)
...
+   * @param delta number of bytes read from the channel in the current read.
    */
-  public ContentWriteCallback(HttpContent httpContent, boolean isLast, ReadIntoCallbackWrapper callbackWrapper) {
-    this.httpContent = httpContent;
-    this.isLast = isLast;
-    this.callbackWrapper = callbackWrapper;
+  private void continueReadIfPossible(long delta) {
+    if (!channel.config().isAutoRead()) {
+      if (bytesBuffered.addAndGet(delta) < bufferWatermark) {
+        channel.read();
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Is `bytesBuffered` being used as intended? Shouldn't it get decremented as data is read or something?

Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Ah, I see that you pass in a negative value when done consuming the content. I think the javadoc needs to be corrected - "the number of bytes read from the channel in the current read" is not accurate. Also, `delta` implies "difference", so if this variable can be negative, I suggest renaming it.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 11]
```diff
@@ -488,91 +523,106 @@ private void validateState(HttpContent httpContent)
...
+     * @param result The result of the request. This would be non null when the request executed successfully
+     * @param exception The exception that was reported on execution of the request
+     */
+    @Override
+    public void onCompletion(Long result, Exception exception) {
+      if (httpContent != null) {
+        ReferenceCountUtil.release(httpContent);
+      }
+      callbackWrapper.updateBytesRead(result);
+      continueReadIfPossible(-result);
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: I assume this call to `continueReadIfPossible()` and its implementation are the only real changes here?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java [hunk 12]
```diff
@@ -434,6 +458,17 @@ protected void writeContent(AsyncWritableChannel writeChannel, ReadIntoCallbackW
   }
 
   /**
+   * Switches auto reading from the channel on and off.
+   * @param autoRead {@code true} if auto read has to be switched on. {@code false} otherwise.
+   */
+  protected void setAutoRead(boolean autoRead) {
+    channel.config().setAutoRead(autoRead);
+    channel.config().setMaxMessagesPerRead(autoRead ? savedMaxMessagesPerRead : 1);
```
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Will sync up with you on exactly what setting maxmessages means  - but do you have to set this before setting autoread to true (as in does setting autoread to true potentially trigger things asynchronously?)

**Comment by pnarayanan** [comment]
Done with a pass of the production code. Will sync up on a couple of things that I am unclear about before I review the tests.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 6 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Thought that would be redundant since POST is the only one that can have content/data. Do you think it is useful?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 8 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: This variable is used in this class for all the watermark checking (`NettyRequest` does not have instance of `NettyConfig`). The variable itself is set in `NettyServer` based on the config. I thought this would the easiest thing to do since this value is constant across all instances of `NettyRequest` and it didn't make sense to pass it through the constructor.

Could you explain why you think this may not be required?

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: There is no practical difference b/w anything <= 0

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 9 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Yes. Specifically this is the number of messages to read per `channel.read()` acc to the [documentation](http://netty.io/4.0/api/io/netty/channel/ChannelConfig.html#setMaxMessagesPerRead(int)). But each message should translate to a single `HttpContent`.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 10 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Will fix the doc.

About `delta`, I have used the same name that is used in `AtomicLong` where this value can be negative. Do you have suggestions for names that might be clearer?

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 11 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Yes, most are formatting changes. I forgot to include that comment here.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 12 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: I can make it the other way round but from what I can gather from the documentation, if autoRead is set to `true`, the channel is read continuously and maxMessagesPerRead should not matter. It only matters when autoRead is `false` and you care about how much you read once you call `channel.read()`.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java [hunk 13]
```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
...
+   * {@code content}.
+   * @param numChunksToAddBeforeRead the number of {@link HttpContent} to add before making the
+   * {@link NettyRequest#readInto(AsyncWritableChannel, Callback)} call.
+   * @throws Exception
+   */
+  private void doBackPressureTest(String digestAlgorithm, ByteBuffer content, List<HttpContent> httpContents,
+      int numChunksToAddBeforeRead)
+      throws Exception {
+    if (numChunksToAddBeforeRead < 0 || numChunksToAddBeforeRead > httpContents.size()) {
+      throw new IllegalArgumentException("Illegal value of numChunksToAddBeforeRead");
```
Comment (pnarayanan) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: this is guarding against the test itself, a practice we don't usually follow - you think this is necessary?

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java [hunk 14]
```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
...
+        queuedReads.set(0);
+        if (future == null && addedCount == numChunksToAddBeforeRead) {
+          future = nettyRequest.readInto(writeChannel, callback);
+        }
+        final HttpContent httpContent = httpContents.get(addedCount);
+        bytesToVerify += (httpContent.content().readableBytes());
+        suspended = bytesToVerify >= NettyRequest.bufferWatermark;
+        addedCount++;
+        nettyRequest.addContent(httpContent);
+        int expectedRefCountOnAdd = future == null || httpContent.content().nioBufferCount() > 0 ? 2 : 1;
```
Comment (pnarayanan) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: could you explain this check?

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java [hunk 15]
```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
...
+    while (addedCount < httpContents.size()) {
+      if (suspended) {
+        assertEquals("There should have been no reads queued when over buffer watermark", 0, queuedReads.get());
+        if (future == null) {
+          future = nettyRequest.readInto(writeChannel, callback);
+        }
+        int chunksRead = readAndVerify(bytesToVerify, writeChannel, content);
+        assertEquals("Number of reads triggered is not as expected", chunksRead, queuedReads.get());
+        // collapse many reads into one
+        queuedReads.set(1);
```
Comment (pnarayanan) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: At this point, nothing is queued, so why is this set to 1?

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java [hunk 16]
```diff
@@ -749,7 +839,140 @@ private void doContentAddAndReadTest(String digestAlgorithm, ByteBuffer content,
...
+        }
+        final HttpContent httpContent = httpContents.get(addedCount);
+        bytesToVerify += (httpContent.content().readableBytes());
+        suspended = bytesToVerify >= NettyRequest.bufferWatermark;
+        addedCount++;
+        nettyRequest.addContent(httpContent);
+        int expectedRefCountOnAdd = future == null || httpContent.content().nioBufferCount() > 0 ? 2 : 1;
+        assertEquals("Reference count is not as expected", expectedRefCountOnAdd, httpContent.refCnt());
+      }
+    }
```
Comment (pnarayanan) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: It seems like you could still cover the case where content gets added regardless of being over the watermark  with yet another flag passed in as an argument - but I realize it can make these checks complicated and may even be outside the scope of what you want to test here.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 12 above)
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Got it.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 10 above)
Comment (pnarayanan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: I guess delta is okay as long as the documentation makes it clearer.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 13 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: This function is meant to be used by multiple tests and multiple cases - so i thought it is useful to have the check here to avoid spurious failures.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 15 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: This collapses many reads on the read channel into one read. This read is for the read/socket channel and not for the write channel (which like you said has nothing queued)

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 14 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: This is a check for ref counts.

If `readInto()` has not been called, `future` will be `null`. In this case, regardless of the nature of the `ByteBuf`, it is retained. Therefore expected `refCnt` is 2 if `future == null`.
If `readInto()` has been called (`future` will be non null), then the `refCnt` increases if the the `ByteBuf` can expose its contents as a `ByteBuffer`. If it cannot and therefore, the contents need to be copied (look at `NettyRequest::writeContent()`), the `refCnt` does not increase.

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java (hunk 16 above)
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyRequestTest.java]: Yeah, I think those checks are out of scope. This test is only meant to test two things
1. `channel.read()` is not called if data watermark has been reached or exceeded.
2. `channel.read()` is called if below data watermark, or if data consumption resulted in data going below the watermark after having exceeded it.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java (hunk 10 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyRequest.java]: Done

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMessageProcessor.java (hunk 7 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMessageProcessor.java]: As discussed offline, this is just to handle non POST multipart requests smoothly if possible.

**Comment by vgkholla** [comment]
Yes built and tested after all changes.
